src/term.rs

```diff
@@ -201,7 +202,157 @@ impl PendingContract {
     }
 }
 
-/// The attributes of an Array.
+pub mod array {
+    use std::mem::transmute;
+    use std::mem::ManuallyDrop;
+
+    use super::*;
+
+    #[derive(Debug, Clone, PartialEq)]
+    pub struct Array {
+        inner: Rc<[RichTerm]>,
+        start: usize,
+        end: usize,
+    }
+
+    impl Array {
+        /// Creates a Nickel array from reference-counted slice.
+        pub fn new(inner: Rc<[RichTerm]>) -> Self {
+            let start = 0;
+            let end = inner.len();
+
+            Self { inner, start, end }
+        }
+
+        /// Returns the effective length of the array.
+        pub fn len(&self) -> usize {
+            self.end - self.start
+        }
+
+        /// Returns `true` if the array is empty.
+        pub fn is_empty(&self) -> bool {
+            self.end == self.start
+        }
+
+        /// Returns a reference to the term at the given index.
+        pub fn get(&self, idx: usize) -> Option<&RichTerm> {
+            self.inner.get(self.start + idx)
+        }
+
+        /// Discards the first `diff` terms of the array.
+        pub fn advance_by(mut self, diff: usize) -> Self {
+            self.start += usize::min(diff, self.len());
+            self
+        }
+
+        /// Makes a mutable slice into the given `Array`.
+        pub fn make_mut(&mut self) -> &mut [RichTerm] {
+            // NOTE: trying to use `Rc::make_mut` will result in the following compiler error:
+            // > the trait `Clone` is not implemented for `[term::RichTerm]`
+            // This seems to be an edge-case in the standard library.
+            // As a workaround, if `get_mut` fails we recollect the array into a new `Rc` by cloning all terms.
+            // Thus we make sure that the second call to `get_mut` won't fail.
+            if Rc::get_mut(&mut self.inner).is_none() {
```

Comment:
Wouldn't a call to something like `Rc::strong_count(&self.inner) != 1` be cheaper? Or at least more natural.

Reply:
After digging, you also need to take into account the weak cout. In the stdlib, `make_mut` use the non public function `Rc::is_unique`, which is defined as:

```rust
    #[inline]
    fn is_unique(this: &Self) -> bool {
        Rc::weak_count(this) == 0 && Rc::strong_count(this) == 1
    }
```